internal/spoke_k8s_client/factory.go

```diff
@@ -1,90 +1,255 @@
 package spoke_k8s_client
 
 import (
+	"context"
+	"fmt"
+	"net/http"
+	"slices"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
-	client "sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 //go:generate mockgen --build_flags=--mod=mod -package=spoke_k8s_client -destination=mock_spoke_k8s_client_factory.go . SpokeK8sClientFactory
 type SpokeK8sClientFactory interface {
-	CreateFromSecret(secret *corev1.Secret) (SpokeK8sClient, error)
-	ClientAndSetFromSecret(secret *corev1.Secret) (SpokeK8sClient, *kubernetes.Clientset, error)
+	CreateFromSecret(ctx context.Context, secret *corev1.Secret) (SpokeK8sClient, error)
+	ClientAndSetFromSecret(ctx context.Context, secret *corev1.Secret) (SpokeK8sClient, *kubernetes.Clientset, error)
+}
+
+// SpokeK8sClientFactoryBuilder contains the data and logic needed to create a spoke client factory. Don't create
+// instances of this directly, use the NewFactory function instead.
+type SpokeK8sClientFactoryBuilder struct {
+	logger            logrus.FieldLogger
+	hubClient         ctrlclient.Client
+	transportWrappers []func(http.RoundTripper) http.RoundTripper
 }
 
 type spokeK8sClientFactory struct {
-	log logrus.FieldLogger
+	logger            logrus.FieldLogger
+	hubClient         ctrlclient.Client
+	transportWrappers []func(http.RoundTripper) http.RoundTripper
+}
+
+// NewFactory creates a builder that can then be used to configure and create a spoke client factory.
+func NewFactory() *SpokeK8sClientFactoryBuilder {
+	return &SpokeK8sClientFactoryBuilder{}
+}
+
+// SetLogger sets the object that will be used to write to the log. This is mandatory.
+func (b *SpokeK8sClientFactoryBuilder) SetLogger(value logrus.FieldLogger) *SpokeK8sClientFactoryBuilder {
+	b.logger = value
+	return b
+}
+
+// SetHubClient sets the client that will be used to call the API of the hub cluster. This is mandatory.
+func (b *SpokeK8sClientFactoryBuilder) SetHubClient(value ctrlclient.Client) *SpokeK8sClientFactoryBuilder {
```

Comment:
why we can't provide client while creating the object? 
we have it here https://github.com/openshift/assisted-service/pull/7090/files#diff-891ba2cfffd82a8ae4131c88beb092d2a88149f579c931e3a1f7ca77fbfc82a5L166 no?

Reply:
sorry my fault missed 
https://github.com/openshift/assisted-service/pull/7090/files#diff-c444f711e9191b53952edb65bfd8c644419fc7695c62611dc0fb304b4fb197d6R625

Reply:
This is just a way to make the code cleaner, avoiding long lists of parameters. We could pass the the logger, the client (and the transport wrapper, only used currently for tests) as parameters to the "New..." function, but over time that results in long lists of parameters like this.

```go
api = NewManager(common.GetTestLog(), db, testing.GetDummyNotificationStream(ctrl), mockEventApi, nil, nil, nil, nil, &config, &leader.DummyElector{}, nil, nil, true, nil, nil, false)
```

It is already useful to avoid the setting the transport wrapper parameter to nil.

Reply:
But this is required params and in this case you left them  as optional so i don't understand actually why it is good

Reply:
I believe it is good for several reasons:

1. It is consistent: all the parameters (required or optional) are provided in the same way.

3. It makes it clearer what each parameter means. Not in this case, but if you had two parameters that are strings it is not the same to see this:

```go
whatever, err := NewWhatever("foo", "bar")
```

Than this:

```go
whatever, err := NewWhatever(().
        SetUserName("foo").
        SetPassword("bar").
        Build()
```

In the first case you have to deep digger to find out what is the meaning of the parameters, and in the second it is explicit.

4. It gives room for documenting each parameter separately: the documentation goes in the "Set..." method of the builder.

5. It simplifies building the object in multiple steps, if needed, for example:

```go
builder := NewWhatever()
builder.SetUserName("foo")
if shouldUsePassword {
        builder.SetPassword("bar")
}
whatever, err := builder.Build()
```

6. It simplifies adding multiple values for the same parameter:

```go
whatever, err := NewWahtever().
        SetUserName("foo").
        SetUserName("foo-alias").
        Build()
```

7. It allows adding new optional parameters without having to change the call sites.

I don't want to bore you with my opinions about this. If you find this unacceptable I will change it.

Reply:
Idk, maybe it just me, I just believe that if parameter is required it should be provided as part of function call another way if someone will write 
`whatever, err := NewWahtever().Build()`
it will pass compilation but will fail on the run an i think better to find such error in compilation.
Though it is my personal opinion

Reply:
Just to be sure, i like your proposition i just don't think it should be that way with required params

Reply:
I understand your point of view, and still think that the benefits outweigh the drawbacks. As that is not the key point of this pull request I am changing it to a plain list of parameters. We can have this discussion another time.